File: java/123_BestTimeToBuyAndSellStockIII.java
class Solution {
    public int maxProfit(int[] prices) {
        int t1Cost = Integer.MAX_VALUE;
        int t1Profit = 0;
        int t2Cost = Integer.MAX_VALUE;
        int t2Profit = 0;
        
        for (int p : prices){
            t1Cost = Math.min(t1Cost, p);
            t1Profit = Math.max(t1Profit, p - t1Cost);
            
            t2Cost = Math.min(t2Cost, p - t1Profit);
            t2Profit = Math.max(t2Profit, p - t2Cost);
        }
        return t2Profit;
            
    }
}

File: swift/123_BestTimeToBuyAndSellStockIII.swift
class Solution {
    func maxProfit(_ prices: [Int]) -> Int {
        var t1Cost: Int = Int.max, t1Profit: Int = 0
        var t2Cost: Int = Int.max, t2Profit: Int = 0
        
        for price in prices {
            t1Cost = min(price, t1Cost)
            t1Profit = max(t1Profit, price - t1Cost)
            
            t2Cost = min(t2Cost, price - t1Profit)
            t2Profit = max(t2Profit, price - t2Cost)
        }
        return t2Profit
    }
}
